### PR TITLE
Laboratory work #2, Talanina Lyubov - 20FPL1

### PR DESCRIPTION
**Pull request by lyutl**

### CODE REVIEW

**Comment by ValeriyaKuznetsova** [comment]
push some changes, please; deadline is coming

don't request a review until you do smth

File: lab_2/main.py [hunk 1]
```diff
@@ -13,7 +14,18 @@ def get_freq_dict(tokens: list) -> dict or None:
     :param tokens: a list of tokens
     :return: a dictionary with frequencies
     """
-    pass
+
+    if not isinstance(tokens, list) or not all(isinstance(i, str) for i in tokens):
+        return None
+    freq_dict = {}
+    for i in tokens:
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:21]: better to use names like word/token/etc

File: lab_2/main.py [hunk 2]
```diff
@@ -24,7 +36,15 @@ def get_language_profiles(texts_corpus: list, language_labels: list) -> dict or
     :param language_labels: a list of given language labels
     :return: a dictionary of dictionaries - language profiles
     """
-    pass
+
+    if (not isinstance(texts_corpus, list) or
+            not isinstance(language_labels, list) or not all(isinstance(i, list) for i in texts_corpus) or
+            not all(isinstance(i, str) for i in language_labels)):
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:42]: what does `all` do?

File: lab_2/main.py [hunk 3]
```diff
@@ -24,7 +36,15 @@ def get_language_profiles(texts_corpus: list, language_labels: list) -> dict or
...
+        return None
+    lang_pr = {}
+    for i in range(len(texts_corpus)):
+        lang_pr[language_labels[i]] = get_freq_dict(texts_corpus[i])
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:46]: imagine that you have a large text. The creation of a freq dictionary can take time. Do you really want to do that on each iteration?

File: lab_2/main.py [hunk 4]
```diff
@@ -33,7 +53,18 @@ def get_language_features(language_profiles: dict) -> list or None:
...
     """
-    pass
+
+    if (not isinstance(language_profiles, dict) or
+            not all(isinstance(key, str) for key in language_profiles.keys()) or
+            not all(isinstance(value, dict) for value in language_profiles.values()) or
+            not language_profiles.items()):
+        return None
+    features = []
+    for i in language_profiles.values():
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:63]: naming

File: lab_2/main.py [hunk 5]
```diff
@@ -43,7 +74,25 @@ def get_text_vector(original_text: list, language_profiles: dict) -> list or Non
...
+            not language_profiles.items()):
+        return None
+    unique_words = get_language_features(language_profiles)
+    text_vector = []
+    for i in unique_words:
+        freq_list = []
+        if i in original_text:
+            for lang in language_profiles.values():
+                if i in lang:
+                    freq_list.append(lang.get(i))
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:91]: there is a way to join lines 91 and 93

File: lab_2/main.py (hunk 2 above)
Comment (lyutl) [reply lab_2/main.py:43]: together with isinstance it checks if all elements in a list(or a dict) are of a certain type

File: lab_2/main.py (hunk 3 above)
Comment (ValeriyaKuznetsova) [reply lab_2/main.py]: not answered

File: lab_2/main.py [hunk 6]
```diff
@@ -83,14 +84,11 @@ def get_text_vector(original_text: list, language_profiles: dict) -> list or Non
...
         freq_list = []
-        if i in original_text:
-            for lang in language_profiles.values():
-                if i in lang:
-                    freq_list.append(lang.get(i))
-        else:
-            freq_list.append(0)
+        for lang in language_profiles.values():
+            if word in lang:
+                freq_list.append(lang.get(word) if word in original_text else 0)
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:91]: there is even a shorter way, check what arguments `get` needs

File: lab_2/main.py [hunk 7]
```diff
@@ -161,7 +170,33 @@ def predict_language_knn(unknown_text_vector: list, known_text_vectors: list,
...
+            not isinstance(k, int) or metric not in ('euclid', 'manhattan')):
+        return None
+    dist_list = []
+    for lang in known_text_vectors:
+        if metric == 'euclid':
+            dist_list.append(calculate_distance(unknown_text_vector, lang))
+        elif metric == 'manhattan':
+            dist_list.append(calculate_distance_manhattan(unknown_text_vector, lang))
+    close_dist = sorted(dist_list)[:k]
+    close_lang = [language_labels[dist_list.index(dist)] for dist in close_dist]
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:192]: what happens here?

File: lab_2/main.py [hunk 8]
```diff
@@ -161,7 +170,33 @@ def predict_language_knn(unknown_text_vector: list, known_text_vectors: list,
...
+    count_lang = {}
+    for lang in close_lang:
+        if lang not in count_lang:
+            count_lang[lang] = 0
+        count_lang[lang] += 1
+    closest_lang = max(count_lang, key=count_lang.get)
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:198]: what happens here?

File: lab_2/main.py [hunk 9]
```diff
@@ -172,7 +207,22 @@ def get_sparse_vector(original_text: list, language_profiles: dict) -> list or N
...
+            not all(isinstance(value, dict) for value in language_profiles.values()) or
+            not language_profiles.items()):
+        return None
+    unique_words = get_language_features(language_profiles)
+    word_scores = []
+    for index, word in enumerate(unique_words):
+        for lang in language_profiles.values():
+            if word in lang:
+                word_scores.append([index, lang.get(word)])
+    sparse_vector = [word_scores[index] for index, word in enumerate(unique_words)
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:223]: what happens here?

File: lab_2/main.py [hunk 10]
```diff
@@ -182,7 +232,21 @@ def calculate_distance_sparse(unknown_text_vector: list,
...
     """
-    pass
+
+    if (not isinstance(unknown_text_vector, list) or
+            not isinstance(known_text_vector, list) or
+            not all(isinstance(i, list) for i in unknown_text_vector) or
+            not all(isinstance(i, list) for i in known_text_vector)):
+        return None
+    unknown_dict = dict(unknown_text_vector)
+    known_dict = dict(known_text_vector)
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:242]: what happens when you use `dict` here?

File: lab_2/main.py [hunk 11]
```diff
@@ -182,7 +232,21 @@ def calculate_distance_sparse(unknown_text_vector: list,
...
+    combine_dict = dict(unknown_text_vector[:])
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:243]: why` [:]`?

File: lab_2/start.py [hunk 12]
```diff
@@ -38,6 +43,31 @@
...
+    known_vectors = [get_sparse_vector(text, lang_profiles) for text in text_corpus]
+
+    for text in UNKNOWN_SAMPLES:
+        unknown_text = remove_stop_words(tokenize(text), STOP_WORDS)
+        unknown_vector = get_sparse_vector(unknown_text, lang_profiles)
+        prediction = predict_language_knn_sparse(unknown_vector, known_vectors, lang_labels, KNN)
+        RESULT.append(prediction[0])
+
     # DO NOT REMOVE NEXT LINE - KEEP IT INTENTIONALLY LAST
     assert RESULT, 'Detection not working'
```
Comment (ValeriyaKuznetsova) [inline lab_2/start.py:73]: RESULT == EXPECTED (put a language in expected)

File: lab_2/start.py [hunk 13]
```diff
@@ -38,6 +43,31 @@
...
+        lang_labels.append('lat')
+
+    lang_profiles = get_language_profiles(text_corpus, lang_labels)
+    known_vectors = [get_sparse_vector(text, lang_profiles) for text in text_corpus]
+
+    for text in UNKNOWN_SAMPLES:
+        unknown_text = remove_stop_words(tokenize(text), STOP_WORDS)
+        unknown_vector = get_sparse_vector(unknown_text, lang_profiles)
+        prediction = predict_language_knn_sparse(unknown_vector, known_vectors, lang_labels, KNN)
+        RESULT.append(prediction[0])
```
Comment (ValeriyaKuznetsova) [inline lab_2/start.py:70]: can you describe in writing what happens in` start.py`?

**Review by ValeriyaKuznetsova — COMMENTED** [review]
Looks good! Some changes are required.

File: lab_2/start.py (hunk 13 above)
Comment (lyutl) [reply lab_2/start.py:70]: imports module os and functions from main.py
creates paths
opens and reads different languages text and samples files
creates empty lists of texts (texts corpus), language labels, and stop_words, and also assigns a value to KNN
then tokenizes all of the texts, removes stop words from them and adds them to the texts corpus as well as adds their language labels to the specific list
creates language profiles of these languages and then gets a sparse vector for each of the texts in the texts corpus
and after that it tokenizes the unknown text, removes stop words from it and also gets its sparse vector
and then predicts what language is the unknown text in by comparing all the sparse vectors

**Pull request closed without merging**